juju/juju
# Expose raft apply metrics

These metrics report exactly the time it takes to apply an operation to
the raft log. Instead of using the client timings, which also factor in
network latency, the following ensures that we just call out the raft
log times. This is important when trying to understand where the cause
of slowness is.

Interestingly, it's actually cheap for us to query if the node state is the
leader. 

## QA steps

```sh
$ juju bootstrap lxd test1 --build-agent --config features="[raft-api-leases,raft-batch-fsm]"
$ juju switch controller
$ juju add-user prometheus
$ juju grant prometheus read controller
$ juju change-user-password prometheus # keep note of password
$ juju show-controller test1 --format=json | jq '.test1.details."api-endpoints"[0]' | xargs -I % echo "https://%/introspection/metrics" | xargs curl -k -s -u user-prometheus:$PASSWORD | grep "raftapply"
```

Should output the following:

```sh
# HELP juju_raftapply_applications Application times for applying to the raft log in ms
# TYPE juju_raftapply_applications summary
juju_raftapply_applications{result="leader-error",quantile="0.5"} 0.121702
juju_raftapply_applications{result="leader-error",quantile="0.9"} 0.167319
juju_raftapply_applications{result="leader-error",quantile="0.99"} 0.167319
juju_raftapply_applications_sum{result="leader-error"} 0.716836
juju_raftapply_applications_count{result="leader-error"} 6
juju_raftapply_applications{result="success",quantile="0.5"} 9.183899
juju_raftapply_applications{result="success",quantile="0.9"} 14.77059
juju_raftapply_applications{result="success",quantile="0.99"} 15.620271
juju_raftapply_applications_sum{result="success"} 115.21207599999998
juju_raftapply_applications_count{result="success"} 13
```
